domain/status/service/service.go

```diff
@@ -265,9 +266,15 @@ func (s *Service) SetApplicationStatusForUnitLeader(
 		return errors.Capture(err)
 	}
 
-	return s.leaderEnsurer.WithLeader(ctx, appName, unitName.String(), func(ctx context.Context) error {
+	err = s.leaderEnsurer.WithLeader(ctx, appName, unitName.String(), func(ctx context.Context) error {
 		return s.st.SetApplicationStatus(ctx, appID, encodedStatus)
 	})
+	if errors.Is(err, corelease.ErrNotHeld) {
+		return statuserrors.UnitNotLeader
```

Comment:
Maybe I'm missing context. This seems strange to check if an error is of X type, and instead of wrapping it or whatever we return a completely different error?

Reply:
Usually we would wrap it, but `corelease.ErrNotHeld` is too much like business logic to raise up. So instead we transform it into a specific, tailor-made domain error

Reply:
We do the same kind of thing in the apiserver, transform domain errors into param/apiserver errors